design_pim/overview.rst

```diff
@@ -0,0 +1,356 @@
+
+Note:
+Since the 1.8 version, we migrated to webpack to manage our front dependencies. To ease the migration we decided to keep the historicals ``requirejs.yml`` files to give the community enough time to migrate their configuration.
+
+Create our first RequireJS module
+=================================
+
+RequireJS modules are the equivalent of php classes. They provide a way to declare dependencies to a module (like use statements) and expose a new module.
+
+Simple business logic
++++++++++++++++++++++
+
+As an example, we will create a module that will display in the console the current UI locale
+
+To begin here is the business logic of our component:
+
+.. code-block:: javascript
+    :linenos:
+
+    {
+        logLocale: function () {
+            console.log(UserContext.get('uiLocale'));
+        }
+    }
+
+
+Our newly created object has one method which log the current catalog locale using the UserContext module.
+Now that we have our business logic implemented, we need to tell the system that our module has a dependency to ``pim/user-context`` and that we are exposing this object.
+
+The real thing
+++++++++++++++
+
+Here is how to do it using RequireJS:
+
+.. code-block:: javascript
+    :linenos:
+
+    // src/Acme/Bundle/CustomBundle/Resources/public/js/locale_logger.js
+    define(
+        ['pim/user-context'],
+        function (UserContext) {
+            return {
+                logLocale: function () {
+                    console.log(UserContext.get('uiLocale'));
+                }
+            }
+        }
+    )
+
+
+The ``define`` method takes two arguments: the first one is an array of the dependencies we want to use (``pim/user-context`` in this case) and a callback that will be called every time our module will be requested. This callback will receive as argument the list dependencies we asked for. Note that you can name the arguments of the callback as you want.
+
+Register it
++++++++++++
+
+Now that we have this small module, we would like to use it somewhere else. To do so, we need to register it in the RequireJS configuration file.
+
+.. code-block:: yaml
+
+    # src/Acme/Bundle/CustomBundle/Resources/config/requirejs.yml
+    config:
+        paths:
+            my_locale_logger: acmecustom/js/locale_logger
+
+
+We just declared a new module named ``my_locale_logger`` located in ``web/bundles/acmecustom/js/locale_logger.js``.
+Note that we don't need to specify the file extension and it is located in the public web folder (moved during the ``bin/console assets:install`` command execution). Also, ``my_locale_logger`` name is a arbitrary name and you can replace it with whatever you want.
+
+Make it available
++++++++++++++++++
+
+To make your module available you need to clear your cache:
+
+.. code-block:: bash
+
+    rm -rf ./var/cache/*
+
+
+dump the assets in the web folder
+
+.. code-block:: bash
+
+    bin/console assets:install web
+
```

Comment:
I think we need to put pim:installer:dump-require-paths after this

Reply:
I never ran this command so maybe it's not mandatory?

Reply:
Oh that's right it gets run in the pim install command